File: packages/borg-sdk/src/dsl.ts
/**
 * The author-side DSL: structs, pipelines, and the repo they make up.
 *
 * This half is pure declaration — nothing here opens a connection or reads a cell. What it produces
 * is a `describe` payload, which is the DSL's compile target and the whole of the contract between
 * an SDK and `borg repo push` (§17.4). A repo that describes itself needs no new engine surface.
 *
 * ## Ownership is explicit, and checked in both directions
 *
 * `borg.bool().derived()` says a pipeline owns the field; `{ writes: ["isInvestible"] }` says which
 * one. Neither is inferred from the other, and assembling the description errors if they disagree —
 * a `derived()` field nobody writes is a field no client may write either and no pipeline ever will,
 * and a `writes` naming a field that is not `derived()` is a write the engine will refuse at the
 * first invocation. Both are static facts, so both are push-time errors rather than runtime ones.
 *
 * Inference is later sugar, and would happen here rather than at runtime.
 */

import type { AnyFieldType, FieldType, ValueOf } from "./values.js";
import type { Description, FieldSpec, StructSpec } from "./protocol.js";

/** A repo the DSL rejected before it ever reached the engine. */
export class BorgDefinitionError extends Error {
  override readonly name = "BorgDefinitionError";
}

export type Fields = Record<string, AnyFieldType>;

export interface StructDef<F extends Fields = Fields> {
  readonly name: string;
  readonly fields: F;
}

/**
 * Declare a struct. Names are used exactly as written, here and on the wire: a `headcount` in the
 * DSL is `Company#1.headcount` at the CLI. No case is converted, because a mapping that is invisible
 * in both directions is a mapping somebody eventually has to reverse-engineer from an error message.
 */
export function struct<F extends Fields>(name: string, fields: F): StructDef<F> {
  if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(name)) {
    throw new BorgDefinitionError(`\`${name}\` is not a usable struct name`);
  }
  if (Object.keys(fields).length === 0) {
    throw new BorgDefinitionError(`struct \`${name}\` declares no fields`);
  }
  return { name, fields };
}

/** The input entity of one invocation. Every access is a wire message; nothing is cached. */
export interface EntityContext<F extends Fields> {
  /** The entity's cell address, as the engine named it: `Company:o-1234abcd`. */
  readonly ref: string;
  /** Read one of the entity's fields. Recorded server-side whether or not it exists. */
  get<K extends keyof F & string>(field: K): Promise<ValueOf<F[K]> | null>;
  /** Write one of the entity's fields. `null` deletes it. */
  set<K extends keyof F & string>(field: K, value: ValueOf<F[K]> | null): Promise<void>;
}

/**
 * Random access to anything else, for the hops a pipeline makes beyond its own entity.
 *
 * Stringly in v1: a cell is named by its text address, and a value by its text form unless a field
 * type is supplied to convert it. Generated types are what will make this typed — the second
 * overload is where they slot in, taking the same `FieldType` the DSL already produces — so the
 * shape is chosen now and only the source of the type changes later.
 */
export interface World {
  get(cell: string): Promise<string | null>;
  get<T>(cell: string, as: FieldType<T>): Promise<T | null>;
  set(cell: string, value: string | null): Promise<void>;
  set<T>(cell: string, value: T | null, as: FieldType<T>): Promise<void>;
}

export type PipelineBody<F extends Fields> = (
  entity: EntityContext<F>,
  world: World,
) => Promise<void>;

export interface PipelineDef<F extends Fields = Fields> {
  readonly name: string;
  readonly source: StructDef<F>;
  /** The fields this pipeline owns. Cross-checked against `derived()` at describe time. */
  readonly writes: readonly string[];
  readonly body: PipelineBody<F>;
}

export interface PipelineOptions<F extends Fields> {
  writes: readonly (keyof F & string)[];
}

/**
 * Declare a pipeline: a producer that maps over one struct, one entity at a time (§4.2).
 *
 * The body is `async` on purpose and verbosely so. Every `get` is a round trip whose result the
 * engine records as a dependency, which is what makes invalidation field-granular; preloading the
 * entity would collapse that read-set to object granularity and cost exactly the property scenario
 * 030 demonstrates. Property-access sugar over `worker_thread` + `Atomics.wait` is the later path,
 * and it does not change any of this.
 */
export function pipeline<F extends Fields>(
  name: string,
  source: StructDef<F>,
  options: PipelineOptions<F>,
  body: PipelineBody<F>,
): PipelineDef<F> {
  if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(name)) {
    throw new BorgDefinitionError(`\`${name}\` is not a usable pipeline name`);
  }
  return { name, source, writes: [...options.writes], body };
}

export interface RepoConfig {
  /**
   * The repo id. Optional, and cross-checked rather than used: `borg.toml` is authoritative,
   * because a repo is a directory and one directory has one id however many modules it contains.
   * Stating it here gets that copy verified at push time instead of quietly ignored.
   */
  id?: number;
  structs: readonly StructDef[];
  pipelines: readonly PipelineDef[];
}

/**
 * Assemble the `describe` payload, refusing anything the engine would refuse later — or worse,
 * would accept and leave unwritable.
 */
export function describe(config: RepoConfig): Description {
  const structs = new Map<string, StructDef>();
  for (const def of config.structs) {
    if (structs.has(def.name)) {
      throw new BorgDefinitionError(`struct \`${def.name}\` is declared twice in this repo`);
    }
    structs.set(def.name, def);
  }

  // Which pipeline claims which field. Built first, because the two cross-checks below are its two
  // directions and both need the whole map.
  const owners = new Map<string, string>();
  const seen = new Set<string>();
  for (const p of config.pipelines) {
    if (seen.has(p.name)) {
      throw new BorgDefinitionError(`pipeline \`${p.name}\` is declared twice in this repo`);
    }
    seen.add(p.name);

    if (structs.get(p.source.name) !== p.source) {
      throw new BorgDefinitionError(
        `pipeline \`${p.name}\` maps over \`${p.source.name}\`, which this repo does not list in ` +
          `\`structs\` — a producer's source struct has to be declared by the repo that implements it`,
      );
    }
    if (p.writes.length === 0) {
      throw new BorgDefinitionError(
        `pipeline \`${p.name}\` declares no \`writes\`, so nothing could ever invoke it`,
      );
    }

    for (const field of p.writes) {
      const what = `${p.source.name}.${field}`;
      const declared = p.source.fields[field];
      if (declared === undefined) {
        throw new BorgDefinitionError(
          `pipeline \`${p.name}\` writes \`${what}\`, which \`${p.source.name}\` does not declare`,
        );
      }
      // Direction one: a claim on a field nobody marked derived. The engine validates every write
      // against declared ownership (§8), so this pipeline would be refused at its first invocation.
      if (!declared.derivedField) {
        throw new BorgDefinitionError(
          `pipeline \`${p.name}\` writes \`${what}\`, which is not declared derived() — ` +
            `add .derived() to the field, or drop it from \`writes\``,
        );
      }
      const already = owners.get(what);
      if (already !== undefined) {
        // Single writer per field is what lets derived layers commit concurrently without
        // conflicting (§16.3). Two claims is not a merge, it is a design mistake.
        throw new BorgDefinitionError(
          `\`${what}\` is written by both \`${already}\` and \`${p.name}\` — a field has one writer`,
        );
      }
      owners.set(what, p.name);
    }
  }

  const specs: StructSpec[] = [];
  for (const def of structs.values()) {
    const fields: FieldSpec[] = [];
    for (const [field, type] of Object.entries(def.fields)) {
      const what = `${def.name}.${field}`;
      const owner = owners.get(what);
      // Direction two: a field declared derived that nothing implements. Clients may not write it
      // (§8) and no producer ever will, so it is a cell that can only ever be empty.
      if (type.derivedField && owner === undefined) {
        throw new BorgDefinitionError(
          `\`${what}\` is declared derived() but no pipeline in this repo writes it — ` +
            `add it to a pipeline's \`writes\`, or drop .derived()`,
        );
      }
      fields.push(
        owner === undefined
          ? { name: field, type: type.wireType }
          : { name: field, type: type.wireType, derived_by: owner },
      );
    }
    specs.push({ name: def.name, fields });
  }

  const description: Description = {
    structs: specs,
    producers: config.pipelines.map((p) => ({ name: p.name, source: p.source.name })),
    // Declared, so the engine knows before it spawns anything that this worker's stdout is its own.
    transport: "socket",
  };
  if (config.id !== undefined) description.repo = config.id;
  return description;
}

File: packages/borg-sdk-py/src/borg/dsl.py
"""
The author-side DSL: structs, pipelines, and the repo they make up.

This half is pure declaration — nothing here opens a connection or reads a cell. What it produces is
a ``describe`` payload, which is the DSL's compile target and the whole of the contract between an
SDK and ``borg repo push`` (§17.4). A repo that describes itself needs no new engine surface.

## Ownership is explicit, and checked in both directions

``borg.bool_().derived()`` says a pipeline owns the field; ``writes=["isInvestible"]`` says which
one. Neither is inferred from the other, and assembling the description errors if they disagree — a
``derived()`` field nobody writes is a field no client may write either and no pipeline ever will,
and a ``writes`` naming a field that is not ``derived()`` is a write the engine will refuse at the
first invocation. Both are static facts, so both are push-time errors rather than runtime ones.

Inference is later sugar, and would happen here rather than at runtime.
"""

from __future__ import annotations

import inspect
import re
from typing import Any, Callable, Iterable, Mapping, Sequence

from .protocol import TRANSPORT
from .values import FieldType

__all__ = [
    "BorgDefinitionError",
    "PipelineDef",
    "StructDef",
    "describe",
    "pipeline",
    "struct",
]

_NAME = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


class BorgDefinitionError(Exception):
    """A repo the DSL rejected before it ever reached the engine."""


class StructDef:
    """A struct, as this repo declares it."""

    __slots__ = ("name", "fields")

    def __init__(self, name: str, fields: Mapping[str, FieldType]) -> None:
        self.name = name
        # Copied so a caller mutating the mapping afterwards cannot change what was described, and
        # ordered because a dict is: `describe` emits fields in the order they were written.
        self.fields = dict(fields)

    def __repr__(self) -> str:
        return f"StructDef({self.name!r}, {list(self.fields)!r})"


def struct(name: str, fields: Mapping[str, FieldType]) -> StructDef:
    """
    Declare a struct. Names are used exactly as written, here and on the wire: a ``headcount`` in the
    DSL is ``Company#1.headcount`` at the CLI. No case is converted — not to ``snake_case`` either,
    which is what a Python SDK would be tempted to do. A mapping that is invisible in both directions
    is one somebody eventually has to reverse-engineer from an error message, and it would also mean
    the same repo pushed from two languages declared two different schemas.
    """
    if not _NAME.match(name):
        raise BorgDefinitionError(f"`{name}` is not a usable struct name")
    if not fields:
        raise BorgDefinitionError(f"struct `{name}` declares no fields")
    for field, declared in fields.items():
        if not isinstance(declared, FieldType):
            raise BorgDefinitionError(
                f"`{name}.{field}` is {declared!r}, not a field type — "
                f"declare it with borg.string(), borg.int_(), and so on"
            )
    return StructDef(name, fields)


class PipelineDef:
    """A producer that maps over one struct, one entity at a time (§4.2)."""

    __slots__ = ("name", "source", "writes", "body", "wants_world")

    def __init__(
        self,
        name: str,
        source: StructDef,
        writes: Sequence[str],
        body: Callable[..., Any],
    ) -> None:
        self.name = name
        self.source = source
        self.writes = tuple(writes)
        self.body = body
        # Decided once, here, rather than on every invocation. A JS body may ignore the second
        # argument by simply not naming it; Python raises, so the arity is read from the signature.
        self.wants_world = _takes_world(body)

    def run(self, entity: Any, world: Any) -> None:
        if self.wants_world:
            self.body(entity, world)
        else:
            self.body(entity)

    def __repr__(self) -> str:
        return f"PipelineDef({self.name!r}, {self.source.name!r}, writes={list(self.writes)!r})"


def _takes_world(body: Callable[..., Any]) -> bool:
    try:
        signature = inspect.signature(body)
    except (TypeError, ValueError):  # a builtin or a C callable: hand it everything
        return True
    positional = 0
    for parameter in signature.parameters.values():
        if parameter.kind is inspect.Parameter.VAR_POSITIONAL:
            return True
        if parameter.kind in (
            inspect.Parameter.POSITIONAL_ONLY,
            inspect.Parameter.POSITIONAL_OR_KEYWORD,
        ):
            positional += 1
    return positional >= 2


def pipeline(
    name: str,
    source: StructDef,
    writes: Sequence[str],
    body: Callable[..., Any] | None = None,
) -> Any:
    """
    Declare a pipeline. Usable as a decorator or as a call::

        @borg.pipeline("invest", Company, writes=["isInvestible"])
        def invest(c, world):
            c.set("isInvestible", c.get("headcount") is not None)

        invest = borg.pipeline("invest", Company, ["isInvestible"], score_it)

    **The body is synchronous, and that is the whole of the difference from the TypeScript SDK.**
    ``c.get("headcount")`` blocks on a socket read and returns the value. Every one is still a round
    trip whose result the engine records as a dependency — which is what makes invalidation
    field-granular — and nothing is preloaded, because preloading would collapse the server-recorded
    read-set to object granularity and cost exactly the property scenario 030 demonstrates. TS spells
    that ``await``; Python spells it nothing at all. The verbosity was never the contract.

    The body may take ``(c)`` or ``(c, world)``; ``world`` is random access to any other cell.
    """
    if not _NAME.match(name):
        raise BorgDefinitionError(f"`{name}` is not a usable pipeline name")

    def define(function: Callable[..., Any]) -> PipelineDef:
        return PipelineDef(name, source, writes, function)

    return define if body is None else define(body)


def describe(
    structs: Iterable[StructDef],
    pipelines: Iterable[PipelineDef],
    id: int | None = None,
) -> dict[str, Any]:
    """
    Assemble the ``describe`` payload, refusing anything the engine would refuse later — or worse,
    would accept and leave unwritable.

    ``id`` shadows the builtin deliberately: it is the keyword an author writes in ``borg.repo(id=1,
    …)``, and matching ``borg.toml``'s spelling matters more here than avoiding the shadow inside one
    function body.
    """
    by_name: dict[str, StructDef] = {}
    for definition in structs:
        if definition.name in by_name:
            raise BorgDefinitionError(
                f"struct `{definition.name}` is declared twice in this repo"
            )
        by_name[definition.name] = definition

    # Which pipeline claims which field. Built first, because the two cross-checks below are its two
    # directions and both need the whole map.
    owners: dict[str, str] = {}
    ordered: list[PipelineDef] = []
    seen: set[str] = set()
    for p in pipelines:
        if p.name in seen:
            raise BorgDefinitionError(f"pipeline `{p.name}` is declared twice in this repo")
        seen.add(p.name)
        ordered.append(p)

        if by_name.get(p.source.name) is not p.source:
            raise BorgDefinitionError(
                f"pipeline `{p.name}` maps over `{p.source.name}`, which this repo does not list "
                f"in `structs` — a producer's source struct has to be declared by the repo that "
                f"implements it"
            )
        if not p.writes:
            raise BorgDefinitionError(
                f"pipeline `{p.name}` declares no `writes`, so nothing could ever invoke it"
            )

        for field in p.writes:
            what = f"{p.source.name}.{field}"
            declared = p.source.fields.get(field)
            if declared is None:
                raise BorgDefinitionError(
                    f"pipeline `{p.name}` writes `{what}`, which `{p.source.name}` does not declare"
                )
            # Direction one: a claim on a field nobody marked derived. The engine validates every
            # write against declared ownership (§8), so this pipeline would be refused at its first
            # invocation.
            if not declared.derived_field:
                raise BorgDefinitionError(
                    f"pipeline `{p.name}` writes `{what}`, which is not declared derived() — "
                    f"add .derived() to the field, or drop it from `writes`"
                )
            already = owners.get(what)
            if already is not None:
                # Single writer per field is what lets derived layers commit concurrently without
                # conflicting (§16.3). Two claims is not a merge, it is a design mistake.
                raise BorgDefinitionError(
                    f"`{what}` is written by both `{already}` and `{p.name}` — "
                    f"a field has one writer"
                )
            owners[what] = p.name

    specs: list[dict[str, Any]] = []
    for definition in by_name.values():
        fields: list[dict[str, Any]] = []
        for field, declared in definition.fields.items():
            what = f"{definition.name}.{field}"
            owner = owners.get(what)
            # Direction two: a field declared derived that nothing implements. Clients may not write
            # it (§8) and no producer ever will, so it is a cell that can only ever be empty.
            if declared.derived_field and owner is None:
                raise BorgDefinitionError(
                    f"`{what}` is declared derived() but no pipeline in this repo writes it — "
                    f"add it to a pipeline's `writes`, or drop .derived()"
                )
            spec: dict[str, Any] = {"name": field, "type": declared.wire_type}
            if owner is not None:
                spec["derived_by"] = owner
            fields.append(spec)
        specs.append({"name": definition.name, "fields": fields})

    description: dict[str, Any] = {
        "structs": specs,
        "producers": [{"name": p.name, "source": p.source.name} for p in ordered],
        # Declared, so the engine knows before it spawns anything that this worker's stdout is its
        # own. Never detected — a detector would have to tell "has not connected yet" from "printed
        # to stdout first", which is the case the socket exists to make harmless (§17.4).
        "transport": TRANSPORT,
    }
    if id is not None:
        description["repo"] = id
    return description
